pkg/combustion/kubernetes.go

```diff
@@ -0,0 +1,149 @@
+	// is usually taking longer to complete due to downloading files
+	log.Audit("Configuring Kubernetes component...")
+
+	configureFunc := kubernetesConfigurator(version)
+	if configureFunc == nil {
+		log.AuditComponentFailed(k8sComponentName)
+		return nil, fmt.Errorf("cannot configure kubernetes version: %s", version)
+	}
+
+	script, err := configureFunc(ctx)
+	if err != nil {
+		log.AuditComponentFailed(k8sComponentName)
+		return nil, fmt.Errorf("configuring kubernetes components: %w", err)
+	}
+
+	log.AuditComponentSuccessful(k8sComponentName)
+	return []string{script}, nil
+}
+
+func kubernetesConfigurator(version string) func(*image.Context) (string, error) {
+	switch {
+	case strings.Contains(version, image.KubernetesDistroRKE2):
+		return configureRKE2
+	case strings.Contains(version, image.KubernetesDistroK3S):
+		return configureK3S
+	default:
+		return nil
+	}
+}
+
+func installKubernetesScript(ctx *image.Context, distribution string) error {
+	sourcePath := "/" // root level of the container image
+	destPath := ctx.CombustionDir
+
+	return ctx.KubernetesScriptInstaller.InstallScript(distribution, sourcePath, destPath)
+}
+
+func configureK3S(_ *image.Context) (string, error) {
+	return "", fmt.Errorf("not implemented yet")
+}
+
+func configureRKE2(ctx *image.Context) (string, error) {
+	if err := installKubernetesScript(ctx, image.KubernetesDistroRKE2); err != nil {
+		return "", fmt.Errorf("copying RKE2 installer script: %w", err)
+	}
+
+	configFile, err := copyKubernetesConfig(ctx, image.KubernetesDistroRKE2)
+	if err != nil {
+		return "", fmt.Errorf("copying RKE2 config: %w", err)
+	}
+
+	installPath, imagesPath, err := ctx.KubernetesArtefactDownloader.DownloadArtefacts(
+		ctx.ImageDefinition.Kubernetes,
+		ctx.ImageDefinition.Image.Arch,
+		ctx.CombustionDir,
+	)
+	if err != nil {
+		return "", fmt.Errorf("downloading RKE2 artefacts: %w", err)
+	}
+
+	rke2 := struct {
+		image.Kubernetes
+		ConfigFile  string
+		InstallPath string
+		ImagesPath  string
+	}{
+		Kubernetes:  ctx.ImageDefinition.Kubernetes,
+		ConfigFile:  configFile,
+		InstallPath: installPath,
+		ImagesPath:  imagesPath,
+	}
+
+	data, err := template.Parse(rke2InstallScript, rke2InstallerScript, &rke2)
+	if err != nil {
+		return "", fmt.Errorf("parsing RKE2 install template: %w", err)
+	}
+
+	installScript := filepath.Join(ctx.CombustionDir, rke2InstallScript)
+	if err = os.WriteFile(installScript, []byte(data), fileio.ExecutablePerms); err != nil {
+		return "", fmt.Errorf("writing RKE2 install script: %w", err)
+	}
+
+	return rke2InstallScript, nil
+}
+
+func copyKubernetesConfig(ctx *image.Context, distro string) (string, error) {
+	if !isComponentConfigured(ctx, k8sConfigDir) {
+		zap.S().Info("Kubernetes config file not provided")
+		return "", nil
+	}
+
+	configDir := generateComponentPath(ctx, k8sConfigDir)
+	configFile := filepath.Join(configDir, k8sConfigFile)
+
+	_, err := os.Stat(configFile)
+	if err != nil {
+		if errors.Is(err, fs.ErrNotExist) {
```

Comment:
Is there a reason you manually used `errors.Is` instead of `os.IsNotExist`? I'm sure it's the same thing under the covers (too lazy to check), but feels safer to rely on the `os` package since that's where `Stat?()` came from in the first place.

Reply:
If you take a look at the documentation of `os.IsNotExist`, it says:
```
// IsNotExist returns a boolean indicating whether the error is known to
// report that a file or directory does not exist. It is satisfied by
// ErrNotExist as well as some syscall errors.
//
// This function predates errors.Is. It only supports errors returned by
// the os package. New code should use errors.Is(err, fs.ErrNotExist).
```

This is also the reason why I might have changed such usages in the past.

Reply:
Good to know :)